pkg/engine/events.go

```diff
@@ -39,35 +39,42 @@ type Event struct {
 	payload interface{}
 }
 
-func NewEvent(typ EventType, payload interface{}) Event {
-	ok := false
-	switch typ {
-	case CancelEvent:
-		ok = payload == nil
-	case StdoutColorEvent:
-		_, ok = payload.(StdoutEventPayload)
-	case DiagEvent:
-		_, ok = payload.(DiagEventPayload)
-	case PreludeEvent:
-		_, ok = payload.(PreludeEventPayload)
-	case SummaryEvent:
-		_, ok = payload.(SummaryEventPayload)
-	case ResourcePreEvent:
-		_, ok = payload.(ResourcePreEventPayload)
-	case ResourceOutputsEvent:
-		_, ok = payload.(ResourceOutputsEventPayload)
-	case ResourceOperationFailed:
-		_, ok = payload.(ResourceOperationFailedPayload)
-	case PolicyViolationEvent:
-		_, ok = payload.(PolicyViolationEventPayload)
-	case PolicyRemediationEvent:
-		_, ok = payload.(PolicyRemediationEventPayload)
-	case PolicyLoadEvent:
-		_, ok = payload.(PolicyLoadEventPayload)
+type EventPayload interface {
+	StdoutEventPayload | DiagEventPayload | PreludeEventPayload | SummaryEventPayload |
+		ResourcePreEventPayload | ResourceOutputsEventPayload | ResourceOperationFailedPayload |
+		PolicyViolationEventPayload | PolicyRemediationEventPayload | PolicyLoadEventPayload
+}
+
+func NewCancelEvent() Event {
```

Comment:
We have a private essentially duplicate of this here: https://github.com/pulumi/pulumi/pull/14590/files#diff-bf13a77fa6342873ea549cba7ee934e312807d7c55de73004a8648de58cddd7dR105-R107.  Might be worth cleaning up.

Reply:
Good spot, will do!